File: 01_GOF23/DP_examples/Python/02_Structural/Flyweight/Flyweight.py
from abc import ABC, abstractmethod

# 形状接口
class IShape(ABC):
    @abstractmethod
    def draw(self):
        pass
    
    @abstractmethod
    def fill(self):
        pass

# 圆形数据
class CircleData:
    def __init__(self, color, radius):
        self.color = color
        self.radius = radius

# 圆形实现
class Circle(IShape):
    def __init__(self, data: CircleData, x: float, y: float):
        self.data = data
        self.x = x
        self.y = y
    
    def draw(self):
        print(f"Draw a Circle with radius {self.data.radius:.2f} at ({self.x:.2f}, {self.y:.2f})")
    
    def fill(self):
        print(f"Fill with {self.data.color}")

File: 01_GOF23/DP_examples/C/02_Structural/Flyweight/flyweight.c
#include "flyweight.h"
#include "unshared_flyweight.h"
#include <stdlib.h>
#include <string.h>

static void Circle_destroy(void* circle) {
    if (!circle) 
        return;
    free(circle);
}

static void CircleData_destroy(void* data) {
    if (!data) 
        return;
    free(data);
}

void Circle_draw(IShape* self) {
    Circle* circle = (Circle*)self;
    printf("Draw a circle with color %s, radius %f at (%f, %f)\n",
           circle->data->color, circle->data->radius,
           circle->x, circle->y);
}

void Circle_fill(IShape* self) {
    Circle* circle = (Circle*)self;
    printf("Fill with %s\n", circle->data->color);
}

IShape* Circle_create_with_data(CircleData* data, float x, float y){
    if (!data) 
        return NULL;
    Circle* circle = (Circle*)malloc(sizeof(Circle));
    if (circle) {
        circle->base.destroy = Circle_destroy;
        circle->base.draw = Circle_draw;
        circle->base.fill = Circle_fill;
        circle->data = data;
        circle->x = x;
        circle->y = y;
        return (IShape*)circle;
    }
    return NULL;
}

IShape* Circle_create(const char* color, float radius, float x, float y) {
    CircleData* data = (CircleData*)malloc(sizeof(CircleData));
    IShape* circle = NULL;
    if (data) {
        data->destroy = CircleData_destroy;
        data->color = color;
        data->radius = radius;
        circle = Circle_create_with_data(data, x, y);
        if (!circle) {
            free(data);
            data = NULL;
        }
    }
    return circle;
}


// 初始化圆形工厂
void CircleFactory_init(CircleFactory* factory) {
    factory->get_circle = Circle_create;
    factory->get_circle_by_data = Circle_create_with_data;
    factory->get_spacial_circle = CircleAddition_create;
}
